File: Data Structures & Algorithms/design-word-search-data-structure/submission-0.py
class TrieNode():
    def __init__(self):
        self.children = {}
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()
        
    def addWord(self, word: str) -> None:
        if not word:
            return

        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isEnd = True

    def search(self, word: str) -> bool:
        return self._dfs(word, 0, self.root)
        

    def _dfs(self, word, index, node) -> bool:
        if index == len(word):
            return node.isEnd
        
        char = word[index]
        
        if char == '.':
            # Try all possible children for wildcard
            for child in node.children.values():
                if self._dfs(word, index + 1, child):
                    return True
            return False
        else:
            # Regular character
            if char not in node.children:
                return False
            return self._dfs(word, index + 1, node.children[char])

File: Data Structures & Algorithms/implement-prefix-tree/submission-0.java
class PrefixTree {
    class TrieNode {
        private TrieNode[] children;
        private boolean isEnd;
        public TrieNode() {
            this.children = new TrieNode[26];
            this.isEnd = false;
        }
    }
    private TrieNode root;
    public PrefixTree() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode node = root;
        for (char w : word.toCharArray()) {
            int index = w - 'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }
            node = node.children[index];
        }
        node.isEnd = true;
    }

    public boolean search(String word) {
        TrieNode node = dfs(word);
        return node != null && node.isEnd;
    }

    public boolean startsWith(String prefix) {
        return dfs(prefix) != null;
    }

    private TrieNode dfs(String str) {
        TrieNode cur = root;
        for (char s : str.toCharArray()) {
            int index = s - 'a';
            if (cur.children[index] == null) {
                return null;
            }
            cur = cur.children[index];
        }
        return cur;
    }
}
